### PR TITLE
support patch container snippet through erda.yml

### PR DESCRIPTION
**Pull request by johnlanni**
#### What type of this PR

Add one of the following kinds:
/kind feature

#### What this PR does / why we need it:
we can use k8s_snippet in the erda.yml like this:

```yaml
version: 2.0
services:
  web:
    ports:
     - port: 8080
    deployments:
      replicas: 1
    resources:
      cpu: 1
      mem: 512
    k8s_snippet:
      container:
        readinessProbe:
          failureThreshold: 20
          periodSeconds: 15
          successThreshold: 1
          tcpSocket:
            port: 8080
          timeoutSeconds: 10
```

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |   erda.yml now support the `k8s_snippet` field which will patch to the container definition of the k8s deployment resource.   |
| 🇨🇳 中文    |   erda.yml 支持了一个新字段 `k8s_snippet`，可以调整对应的 k8s deployment 中关于 container 的配置定义  |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkg/parser/diceyml/fieldname_validate_visitor.go [hunk 1]
```diff
@@ -214,6 +214,23 @@ func (o *FieldnameValidateVisitor) VisitExecCheck(v DiceYmlVisitor, obj *ExecChe
...
 
+func (o *FieldnameValidateVisitor) VisitK8SSnippet(v DiceYmlVisitor, obj *ContainerSnippet) {
+	res, ok := o.currentService["k8s_snippet"].(map[interface{}]interface{})
+	if !ok {
+		return
+	}
+	for k := range res {
+		switch i := k.(type) {
+		case string:
+			if !contain(i, []string{"workingDir", "envFrom", "env", "livenessProbe", "readinessProbe", "startupProbe", "lifeCycle", "terminationMessagePath", "terminationMessagePolicy", "imagePullPolicy", "securityContext", "stdin", "stdinOnce", "tty"}) {
```
Comment (dspo) [inline pkg/parser/diceyml/fieldname_validate_visitor.go:225]: 是不是 `if !contain(k, []string{...` 而不是 `if !contain(i, []string{...`

**Comment by Effet** [comment]
what is the difference with health_check?

File: apistructs/service.go [hunk 2]
```diff
@@ -291,6 +291,8 @@ type Service struct {
 
 	// ProjectServiceName means use service name with servicegroup id when create k8s service
 	ProjectServiceName string `json:"projectServiceName,omitempty"`
+	// K8s Container Snippet
+	K8SSnippet *diceyml.ContainerSnippet `json:"k8sSnippet,omitempty"`
```
Comment (Muzry) [inline apistructs/service.go:295]: Consider reducing the scope of the field `K8SSnippet`

**Comment by johnlanni** [comment]
> what is the difference with health_check?

not only the probes, we can support more k8s native config ability like securyContext and so on.

**Comment by Effet** [comment]
> more k8s native config ability

I'm afraid this is a pandora box that will obliterate the style of `erda`, I suggest we encapsulate k8s' ability more tightly.

**Review by Muzry — COMMENTED** [review]
/lgtm

**Comment by johnlanni** [comment]
> > more k8s native config ability
> 
> I'm afraid this is a pandora box that will obliterate the style of `erda`, I suggest we encapsulate k8s' ability more tightly.

I don't think so.  In this way, the commonly used fields in erda.yml can be kept as simple as possible, and for the in-depth use of k8s, directly use the native configuration snippet of k8s, and do not repeat the definition, which is meaningless.

**Review by Effet — APPROVED** [review]

**Pull request merged**